Easy 1058 cheksums

jo-pol opened the pull request:
fixes a part of EASY-1058
#### When applied it will
- copy the sha1 values from `manifest-sha1.txt` in a bag into the `fo.xml` for file items
- switch on sha1 for file items not found in `manifest-sha1.txt`
- fail on `manifest-sha1.txt` lines with just one field or more than two fields (fields are separated by any sequence of white space)
#### Where should the reviewer @DANS-KNAW/easy start?
- `EasyStageDatasetSpec.run` documents that sha1 values are copied
- `EasyStageDataset.createFileAndFolderSdos` reads and process the `manifest-sha1.txt`
#### How should this be manually tested?
- `mvn clean install` of the project
- feed the created `target/test/sdoPuddings` one by one to `easy-ingest-flow` or `easy-ingest`, an example for the latter:
  
  ```
  cp easy-stage-dataset/target/sdoPuddings easy-dtap/shared # ignored by git
  vagrant ssh
  easy-ingest /vagrant/shared/sdoPuddings/one-invalid-sha1/
  ```
- the example will complain about the cheksum of the third file item (_a value somehow obtained via easy-split-multideposit_)
- examine the other files with http://deasy.dans.knaw.nl:8080/fedora/objects/easy-file:NNN/objectXML NNN is reported by the error message of the ingest, the sha-1will be the same as in the manifest-sha1.txt of the test bag in stage dataset
- try another bag without sha1's, fedora will nevertheless show a sha1
#### related pull requests on github

| repo | PR |
| --- | --- |
| easy- | [PR#](PRlink) |

jo-pol committed:
rename UserCategory -> FileAccessRights

afterburner of EASY-1065-staged-file-rights 7c7093d47dc67fe234fed64d5def2c86df1124c
was postponed to avoid merge conflicts
(cherry picked from commit 09e16a5)

jo-pol committed:
new test bag with SHA-1 check sums

jo-pol committed:
copy checksum from bag into fo.xml, if missing just switch on the checksum in fedora

jo-pol committed:
one more unit test

jo-pol committed:
no logging during build

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -79,6 +81,15 @@ object EasyStageDataset {
+      readFileToString(new File(s.bagitDir, sha1File))
+        .split("\\v+") // split into lines
+        .map(_.split("\\h+")) // split into tokens
+        .filter(!_.isEmpty) // skip empty lines
```
Either use `.filterNot(_.isEmpty)` or `.filter(_.nonEmpty)`.
Personally I prefer the latter...

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/lib/FOXML.scala:
```diff
@@ -32,15 +32,18 @@ object FOXML {
         <dc:title>{label}</dc:title><dc:type>{mimeType}</dc:type>
       </oai_dc:dc>
-    getFOXML(label, ownerId, dc).mkString
+    sha1 match {
```
Please use `map` and `getOrElse` here.

jo-pol committed:
suggested refactoring + run returns emd

returning emd prepares for integration of license generator in easy-ingest-flow
thus reading again is not required

jo-pol committed:
run returns not only emd but also amd

returning emd/amd allows easy-ingest-flow to pass them on to easy-ingest

janvanmansum commented on src/test/scala/nl/knaw/dans/easy/stage/EasyStageDatasetSpec.scala:
```diff
@@ -59,6 +57,22 @@ class EasyStageDatasetSpec extends FlatSpec with Matchers {
     deleteDirectory(bagitDir)
   }
 
+  it should "stumble over a manifest.sha1 with to many fields on a line" in {
```
to -> too

janvanmansum commented on src/test/scala/nl/knaw/dans/easy/stage/EasyStageDatasetSpec.scala:
```diff
@@ -69,34 +83,36 @@ class EasyStageDatasetSpec extends FlatSpec with Matchers {
     createFileAndFolderSdos(dataDir, DATASET_SDO, OPEN_ACCESS_FOR_REGISTERED_USERS) shouldBe a[Success[_]]
-    FileUtils.readFileToString(fileMetadataFile) should include ("<visibleTo>KNOWN</visibleTo>")
-    FileUtils.readFileToString(fileMetadataFile) should include ("<accessibleTo>KNOWN</accessibleTo>")
+    readFileToString(fileMetadataFile) should include ("<visibleTo>KNOWN</visibleTo>")
```
Any specific reason why you don't like the `FileUtils` class to be explicitly here and do use it in the test above?

jo-pol replied:
Just forgot to apply static import occasionally, I guess

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -44,9 +45,9 @@ object EasyStageDataset {
   }
 
-  def run(implicit s: Settings): Try[Unit] = {
+  def run(implicit s: Settings): Try[(EasyMetadata, String)] = {
```
Where are we going to use the `(EasyMetadata, String)` tuple?

jo-pol replied:
As the commit message says: to be passed on by easy-ingest-flow to easy-ingest

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -72,13 +73,22 @@ object EasyStageDataset {
+    val maybeSha1Map: Try[Map[String, String]] = Try {
+      val sha1File = "manifest-sha1.txt"
+      readFileToString(new File(s.bagitDir, sha1File))
+        .split("\\v+") // split into lines
```
Neat one, didn't know that.

rvanheest replied:
you could also use `.lines` here...

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -72,13 +73,22 @@ object EasyStageDataset {
   }
 
   def createFileAndFolderSdos(dir: File, parentSDO: String, rights: AccessCategory)(implicit s: Settings): Try[Unit] = {
+    val maybeSha1Map: Try[Map[String, String]] = Try {
```
Nice, but not really necessary, as have presumably already checked that the bag is valid. If one of the manifests were not correctly formatted that would have shown up. Also, instead of parsing the manifest yourself you could have used the bagit-lib from Library of Congress.

jo-pol replied:
That presumably is not yet implemented, and if it gets implemented in another command we still have to keep the barbarians at the gate here. 

Didn't know about the bagit-lib.

jo-pol replied:
According to @rvanheest that library is quite complex to use and we don't have it on board in this component for any other purposes.

rvanheest replied:
I'd say keep it for now without bagit. Once we start using that more in the TLA project we can always refactor this one as well. Also we're close to a deadline, so I think this hack fix is fine for now.

janvanmansum committed:
Amendments

jo-pol committed:
Merge pull request #4 from janvanmansum/EASY-1058-cheksums

Amendments

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -72,13 +73,22 @@ object EasyStageDataset {
+        .split("\\v+") // split into lines
+        .map(_.split("\\h+")) // split into tokens
+        .filter(_.nonEmpty) // skip empty lines
+        .map(a => if (a.length == 2 && !a(0).matches("[a-zA-Z0-9]")) a(1) -> a(0) else throw new Exception(s"Invalid line in $sha1File: ${a.mkString(" ")}"))
```
Since you expect an `Array` with two elements here, you can do this with a pattern match inside the function as well. That's a little more readable IMO:

``` scala
.map {
  case Array(x, y) if !x.matches("[a-zA-Z0-9]") => y -> x
  case array => throw new IllegalArgumentException(s"Invalid line in $sha1File: ${array.mkString(" ")}")
}
```

Note that you could also discard the second match, in which case the pattern match will automatically throw an exception, although then it is less clear where that exception comes from and you have less/no control over which type of exception you throw or what the message is... That's why I left this second case in.

jo-pol committed:
type alias and pattern matching

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -72,13 +74,23 @@ object EasyStageDataset {
+        .lines.filter(_.nonEmpty)
+        .map(_.split("\\h+")) // split into tokens on sequences of horizontal white space characters
+        .map {
+          case Array(sha1, filePath) if !sha1.matches("[a-zA-Z0-9]") => filePath -> sha1
```
Should be `a-f` etc instead of `a-z`.

jo-pol committed:
type alias and pattern matching

jo-pol commented:
last commit message should have been: pattern matching to validate sha1
